MAJOR feat(fabric) remove callbacks in for Promise support

asturur opened the pull request:
closes #3684

This is not ready yet, all tests are broken for sure, probably the code is broken too.
I gave a first pass at swapping out callbacks for promises but there is still a lot to do.

asturur commented on src/shapes/group.class.js:
```diff
@@ -559,27 +559,14 @@
     delete options.objects;
-    if (typeof objects === 'string') {
-      // it has to be an url or something went wrong.
-      fabric.loadSVGFromURL(objects, function (elements) {
```
this feature has to go away.

ShaMan123 replied:
I think this means you can remove `groupSVGElements`

asturur commented on src/shapes/image.class.js:
```diff
@@ -700,26 +700,25 @@
+            fabric.util.enlivenObjectEnlivables(object, object, function () {
+              var image = new fabric.Image(img, object);
+              callback(image, false);
+            });
```
filter loading needs to be refactored as an enlivabled

asturur commented on src/shapes/itext.class.js:
```diff
@@ -1,14 +1,4 @@
-      object.textDecoration.indexOf('overline') > -1 && (object.overline = true);
-      delete object.textDecoration;
-    }
-  }
```
this was meant to aid for changes between fabric 1.x and fabric 2.0
I think it can go away now.

asturur commented on src/shapes/line.class.js:
```diff
@@ -285,15 +285,15 @@
-    };
+  fabric.Line.fromObject = function(object) {
     var options = clone(object, true);
     options.points = [object.x1, object.y1, object.x2, object.y2];
```
not sure why we export with x1,x2... and we import with a points array, but we need to standardize this.

melchiar replied:
For that matter you could probably make a good argument for removing circle or even triangle as well since we have ellipse and polygon.

ShaMan123 replied:
what's the decision regarding the export?

asturur replied:
I'm not sure where did i write it, but in theory, apart circle, everything else can be done at the same speed with fabric.Path, you can still have constructors for Triangle, but will create a Path.
Then you need to write setters/getters to change the path when you change width/height. Same for line.

This PR won't take care of those things, but we can have a session where we look at those smaller classes and we decided if we want to normalize them.

asturur commented on src/shapes/path.class.js:
```diff
@@ -335,20 +335,10 @@
-  fabric.Path.fromObject = function(object, callback) {
-    if (typeof object.sourcePath === 'string') {
-      var pathUrl = object.sourcePath;
-      fabric.loadSVGFromURL(pathUrl, function (elements) {
```
also this feature, away.
This is a reminescensce of pathGroup, in which if you exported to a data-less object, the path would be converted in a path-string or an svg.

asturur commented on src/shapes/text.class.js:
```diff
@@ -1699,22 +1699,10 @@
-      else {
-        callback(textInstance);
-      }
-    }, 'text');
```
in theory adding 'path' to enlivable should do the trick no?

ShaMan123 replied:
I think so

asturur commented on src/util/lang_object.js:
```diff
@@ -63,7 +63,7 @@
   //TODO: this function return an empty object if you try to clone null
   function clone(object, deep) {
-    return extend({ }, object, deep);
+    return deep ? extend({ }, object, deep) : Object.assign({}, object);
```
since we removed support for ie11 and we plan to add a transpiler, i think we can leverage Object.assign for the shallow clone.

asturur commented on src/util/misc.js:
```diff
@@ -457,55 +457,32 @@
+          img.crossOrigin = options.crossOrigin;
+          img.src = url;
+        }
+      });
```
crossOrigin is no more optional, is necessary everywhere nowadays, we can stop assuming is optional.

asturur commented on src/util/misc.js:
```diff
@@ -538,105 +515,38 @@
-          onLoaded();
+    enlivenObjects: function(objects, namespace, reviver) {
+      return Promise.all(objects.map(function(obj) {
+        var klass = fabric.util.getKlass(obj.type, namespace);
```
if we find a nice solution for filters, namespace can go away

ShaMan123 replied:
same here regarding passing `klass` instead of namespace

asturur commented:
@ShaMan123 @melchiar When i know it actually runs i will ask you to give a look.
For now is wasted time

asturur commented on .eslintrc.json:
```diff
@@ -3,7 +3,7 @@
     "browser": true
   },
   "globals": {
-    "ActiveXObject": true,
```
If you know what is this, you are probably 35+

andrewzolotukhin replied:
confirming. almost 36 and I'm aware of it :)

ShaMan123 commented:
Holy shit this is major!
So much garbage removed! Exciting!

asturur commented on src/mixins/canvas_serialization.mixin.js:
```diff
@@ -31,38 +30,36 @@ fabric.util.object.extend(fabric.StaticCanvas.prototype, /** @lends fabric.Stati
+          backgroundColor: serialized.background,
+          overlayImage: serialized.overlayImage,
+          overlayColor: serialized.overlay,
+          clipPath: serialized.clipPath,
```
as soon as we have es6 in place this will look leaner.

asturur commented on src/util/misc.js:
```diff
@@ -538,104 +515,56 @@
+        return keys.reduce(function(acc, element, index) {
+          acc[keys[index]] = enlived[index];
+          return acc;
+        }, {});
```
@ShaMan123  i changed your work on enlivenObjectEnlivables.
I wanted it to take care of patterns/gradients/
Can you tell me what you think?

ShaMan123 replied:
We can add `objects` as well

asturur replied:
ok we find a way to do that

asturur commented:
There is a little bit of changes here.
Apart promises, things should work in this way:
When you create an object with `var myThing = new fabric.MyThing(options)`,  i wish the standard was that you pass in the options values that are ready to use, and not the serialized part of it.
Serialized things are async and i don't like async constructors or with callback. The general consensus is that constructor shouldn't do much work.

So enliveObjectEnlivables can eventually wrap the options before you create an object, or you start from the fromObject method directly.

Indeed i removed the setBackgroundImage, setOverlayImage.
if you have a source of an image:

`fabric.Image.FromUrl(src).then((fabricImage) => canvas.backgroundImage = fabricImage)`

From a developer persepctive seems reasonable to write something like that.

ShaMan123 reviewed (changes requested):
This PR looks awesome.
Makes the codebase much more readable and elegant and easy to use and override.
I have added comments.
The main things are:
1. keeping customization in mind (e.g. introducing a param/mechanism to allow adding custom enlivened props with ease) 
2. drop namespaces altogether
3. be consist in dropping callbacks - basically drop `_fromObject` as well

ShaMan123 commented on src/shapes/object.class.js:
```diff
@@ -1981,23 +1972,18 @@
    */
   fabric.Object.ENLIVEN_PROPS = ['clipPath'];
```
drop this in favor of something better

ShaMan123 commented on src/shapes/object.class.js:
```diff
@@ -1981,23 +1972,18 @@
-        callback && callback(instance);
-      });
+  fabric.Object._fromObject = function(className, object, extraParam) {
+    var klass = fabric[className], serializedObject = clone(object, true);
```
+1 for dropping fabric.Object._fromObject and letting ` fabric.util.enlivenObjectEnlivables` clone the arg if needed

asturur commented on src/mixins/canvas_serialization.mixin.js:
```diff
@@ -31,38 +30,36 @@ fabric.util.object.extend(fabric.StaticCanvas.prototype, /** @lends fabric.Stati
+    return fabric.util.enlivenObjects(serialized.objects || [], '', reviver)
+      .then(function(enlived) {
+        _this.clear();
+        return fabric.util.enlivenObjectEnlivables({
```
We could eventually try to enlive every single prop, and not having a list.

ShaMan123 replied:
Well if it's possible and not damaging... sound not too bad

asturur replied:
There isn't a real decision. For now a generic enlive all works, as long as you remove the resizeFilter ( thing i did in the image class ). I would keep it as it is, and we circle back when issues surface.
Compared to the previous canvas enliving it seems a step up anyway

asturur commented:
> be consist in dropping callbacks - basically drop _fromObject as well
What do you mean with dropping _fromObject?

ShaMan123 commented:
> > be consist in dropping callbacks - basically drop _fromObject as well
> > What do you mean with dropping _fromObject?

I mean I think this it is redundant - it doesn't do much but wrap the call to `fabric.util.enlivenObjectEnlivables`

asturur commented:
> > > be consist in dropping callbacks - basically drop _fromObject as well
> > > What do you mean with dropping _fromObject?
> 
> I think it is redundant - it doesn't do much but wrap the call to `fabric.util.enlivenObjectEnlivables` Perhaps rename it as something like `prepareObject`

Well if it wraps if for 6-7 classes, is good enough.
When switching to es6 or ts, most of those things shrink in size considerably.

asturur commented:
So the PR would have passing tests, but something is making it stuck in the browser

asturur commented:
@ShaMan123 help me:
```
not ok 30 Firefox 96.0 - [108 ms] - fabric.util.animate: animate with list of values
    ---
        actual: >
            null
        stack: >
            test@http://localhost:8080/node_modules/qunit/qunit/qunit.js:2900:12
            @http://localhost:8080/6514[49](https://github.com/fabricjs/fabric.js/runs/5183084137?check_suite_focus=true#step:6:49)99684/test/unit/animation.js:335:9
            @http://localhost:8080/6[51](https://github.com/fabricjs/fabric.js/runs/5183084137?check_suite_focus=true#step:6:51)44999684/test/unit/animation.js:845:3
            
        message: >
            Expected [52](https://github.com/fabricjs/fabric.js/runs/5183084137?check_suite_focus=true#step:6:52) assertions, but [59](https://github.com/fabricjs/fabric.js/runs/5183084137?check_suite_focus=true#step:6:59) were run
        negative: >
            false
        browser log: |
    ...
```
i swear i looked for this test in all my hard disk and i can't locate it

asturur commented:
ok somehow the test was in latest master, but still running here.
Misteries of the CI.

asturur commented:
ok 55 files changed is no joke to review.
I would suggest this process:
- review the tests first, that will give you an idea of what we broke.
- review the code changes and check if there is something that you necessarly want changed now, and that for some reason can't be an follow up.

asturur commented:
After this PR i would shift my focus onboarding on the group changes

ShaMan123 commented:
> @ShaMan123 help me:
> 
> ```
> not ok 30 Firefox 96.0 - [108 ms] - fabric.util.animate: animate with list of values
>     ---
>         actual: >
>             null
>         stack: >
>             test@http://localhost:8080/node_modules/qunit/qunit/qunit.js:2900:12
>             @http://localhost:8080/6514[49](https://github.com/fabricjs/fabric.js/runs/5183084137?check_suite_focus=true#step:6:49)99684/test/unit/animation.js:335:9
>             @http://localhost:8080/6[51](https://github.com/fabricjs/fabric.js/runs/5183084137?check_suite_focus=true#step:6:51)44999684/test/unit/animation.js:845:3
>             
>         message: >
>             Expected [52](https://github.com/fabricjs/fabric.js/runs/5183084137?check_suite_focus=true#step:6:52) assertions, but [59](https://github.com/fabricjs/fabric.js/runs/5183084137?check_suite_focus=true#step:6:59) were run
>         negative: >
>             false
>         browser log: |
>     ...
> ```
> 
> i swear i looked for this test in all my hard disk and i can't locate it
> 
> When i go to `animation.js:335:9` i have the other test, the one that says animate with abort

@rockerBOO take a look 
Running on windows - might be the animation frame rate is different? seems to lack 2 calls to `onChange` which are 2*7 assertions
@asturur is this possible?
```bash
not ok 30 fabric.util.animate > animate with list of values
  ---
  message: Expected 52 assertions, but 38 were run
  severity: failed
  actual  : null
  expected: undefined
  stack: |
        at Object.test (C:\Users\DELL\Desktop\DEV\fabric.js\node_modules\qunit\qunit\qunit.js:2900:5)
        at C:\Users\DELL\Desktop\DEV\fabric.js\test\unit\animation.js:335:9
        at Object.<anonymous> (C:\Users\DELL\Desktop\DEV\fabric.js\test\unit\animation.js:845:3)
        at Module._compile (internal/modules/cjs/loader.js:1072:14)
        at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
        at Module.load (internal/modules/cjs/loader.js:937:32)
        at Function.Module._load (internal/modules/cjs/loader.js:778:12)
        at Module.require (internal/modules/cjs/loader.js:961:19)
  ...
```

_Originally posted by @ShaMan123 in https://github.com/fabricjs/fabric.js/issues/7633#issuecomment-1035963544_

ShaMan123 commented:
we should simply remove the `assert.expect`

asturur commented:
i removed that because i did not want to deal with it.
I'm sure on browser it can be different because it can skip a frame for whatever reason. also browser run in the command line do not have a real screen connected, the refresh rate is an artifact. ( on my 144hz screen the refresh rate is 144 fps )

ShaMan123 commented on src/shapes/image.class.js:
```diff
@@ -756,8 +739,10 @@
-      extend((options ? fabric.util.object.clone(options) : { }), parsedAttributes));
+    fabric.Image.fromURL(parsedAttributes['xlink:href'], Object.assign({ }, options || { }, parsedAttributes))
+      .then(function(fabricImage) {
+        callback(fabricImage);
```
Is the remaining callback intentional?
Why `fromElement` is left as legacy?

ShaMan123 commented on src/shapes/object.class.js:
```diff
@@ -815,10 +815,6 @@
      */
     setOptions: function(options) {
       this._setOptions(options);
```
`setOptions` doesn't do anything anymore,
maybe rename `_setOptions`?

ShaMan123 commented on src/shapes/object.class.js:
```diff
@@ -1980,25 +1971,19 @@
-        callback && callback(instance);
-      });
+  fabric.Object._fromObject = function(className, object, extraParam) {
+    var klass = fabric[className], serializedObject = clone(object, true);
```
I think we should get rid of `className` and pass a live class (`klass`)

asturur replied:
yes we can, but that is not fundamental for moving this forward. this is ok as an action item

ShaMan123 replied:
alright

ShaMan123 commented on src/util/misc.js:
```diff
@@ -457,55 +457,32 @@
+          resolve(img);
+        };
+        if (!url) {
+          done();
```
isn't it better rejecting in this case?
I understand this is how it was but it's wrong.

asturur replied:
Yes but i don't want to change it now. We can change it , no issues from me

ShaMan123 replied:
not in this PR you mean?

ShaMan123 commented on src/util/misc.js:
```diff
@@ -457,55 +457,32 @@
+        else {
+          img.onload = done;
+          img.onerror = function () {
+            reject(new Error('Error loading ' + img.src));
```
we should pass the original error event
```diff
-            img.onerror = function () {
-            reject(new Error('Error loading ' + img.src));
+            img.onerror = reject;
```

asturur replied:
https://developer.mozilla.org/en-US/docs/Web/API/ErrorEvent

So the ErrorEvent  seems to be different from `new Error()` so we can't get 1:1 mapping reject and onerrorr

ShaMan123 replied:
strange, it seems it has not important data
```txt
Event {isTrusted: true, type: 'error', target: img, currentTarget: img, eventPhase: 2, …}
isTrusted: true
bubbles: false
cancelBubble: false
cancelable: false
composed: false
currentTarget: null
defaultPrevented: false
eventPhase: 0
path: [img]
returnValue: true
srcElement: null
target: null
timeStamp: 382917.6999999881
type: "error"
```
so ok we leave it as is

ShaMan123 replied:
The red log is what I meant important but it isn't part of the error event
![image](https://user-images.githubusercontent.com/34343793/153961779-80704d95-3187-4d32-baa1-5a9895d3f631.png)

ShaMan123 replied:
already it is much better, an error has been raised in a demo I work with that never did and should have

ShaMan123 commented on src/util/misc.js:
```diff
@@ -538,104 +515,52 @@
-      function onLoaded() {
-        if (++numLoadedPatterns === numPatterns) {
-          callback && callback(enlivenedPatterns);
+    enlivenObjectEnlivables: function (serializedObject) {
```
this seems a bit loose and error prone.
shouldn't we pass a map with keys and types?
e.g. 
```js
{
   clipPath: 'object',
   fill: 'pattern',
   path: 'object'
}
```

asturur replied:
As soon as it errors, we can find a solution.
You are not required to pass all in, but you can.
With deconstruction this becomes super easy to do.

ShaMan123 reviewed (changes requested):
I reviewed and here are my findings:
1. namespaces - can we let them go and pass real classes?
2. canvas_serialization - did you miss these?
https://github.com/fabricjs/fabric.js/blob/fee85470b80127b43dc9e1174f4a17d0bd7951be/src/mixins/canvas_serialization.mixin.js#L99
https://github.com/fabricjs/fabric.js/blob/fee85470b80127b43dc9e1174f4a17d0bd7951be/src/mixins/canvas_serialization.mixin.js#L111
I see `test/visual/z_svg_export.js` `test/visual/toDataURL.js` use callbacks instead of promises as well
3. left many comments to change stuff
4. `misc/groupSVGElements` is dead and should be removed?
5. actually I see there are a few dead methods in misc
6. I will try to push some minor doc commits
7. what about svg parsing? I see it remained in callback form `Object.fromElement` and `loadSvgFrom..`
8. you missed `fabric.Image.filters.BlendImage.fromObject`

asturur commented on src/mixins/canvas_serialization.mixin.js:
```diff
@@ -205,24 +132,20 @@ fabric.util.object.extend(fabric.StaticCanvas.prototype, /** @lends fabric.Stati
-    else {
-      callback && callback(clone);
+      // TODO verify this
+      clone.setBackgroundImage(this.backgroundImage.src);
```
this should error, so we miss a test.

asturur commented on src/shapes/group.class.js:
```diff
@@ -559,26 +559,15 @@
-        callback && callback(new fabric.Group(enlivenedObjects, options, true));
+    return fabric.util.enlivenObjects(objects).then(function (enlivenedObjects) {
+      return fabric.util.enlivenObjectEnlivables(options).then(function(enlivedProps) {
+        return new fabric.Group(enlivenedObjects, Object.assign(options, enlivedProps), true);
```
no objects no for now.
It would required doing things recursively, i m fine with enlving the objects manually for groups and canvas right now.
The extra terse code is not giving us much

ShaMan123 replied:
group is phasing out so I don't care

asturur commented:
![image](https://user-images.githubusercontent.com/1194048/154021778-ea068c23-91dc-4b41-88a4-63e2e91af787.png)

@kangax 

I'll write the test for you

asturur commented:
Ok @ShaMan123 @melchiar i think is mergeable now.
- added tests for Vibrance filter and BlendColor filters that were missing
- Fixing @ShaMan123 comments on: using the class instead of classname, blendImage fix, some remaining callbacks
- removed some more dead methods

SVG parsing will be done in a followup.
I'm not even sure i want to change it before evaluating if i should work on a different parser entirely.

ShaMan123 commented:
I have a commit for eraser
I want to push it


בתאריך יום ג׳, 15 בפבר׳ 2022, 13:18, מאת Andrea Bogazzi ‏<
***@***.***>:

> Ok @ShaMan123 <https://github.com/ShaMan123> @melchiar
> <https://github.com/melchiar> i think is mergeable now.
>
>    - added tests for Vibrance filter and BlendColor filters that were
>    missing
>    - Fixing @ShaMan123 <https://github.com/ShaMan123> comments on: using
>    the class instead of classname, blendImage fix, some remaining callbacks
>    - removed some more dead methods
>
> SVG parsing will be done in a followup.
> I'm not even sure i want to change it before evaluating if i should work
> on a different parser entirely.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fabricjs/fabric.js/pull/7657#issuecomment-1040152807>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIGAW4JCMM5L7O2TJWFHJQDU3IZCBANCNFSM5NVQVJBA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

asturur commented:
on a PR on this branch and let's merge in on this

asturur commented:
oh well i see it depends also on the group work.
Well as soon as this approved i ll merge it.
Then i ll get to group rewrite understanding/commenting/collaborating.

ShaMan123 commented:
I want to makr incremental change only for promise and commit that here
And then handle the group chsnge

בתאריך יום ג׳, 15 בפבר׳ 2022, 14:38, מאת Andrea Bogazzi ‏<
***@***.***>:

> oh well i see it depends also on the group work.
> Well as soon as this approved i ll merge it.
> Then i ll get to group rewrite understanding/commenting/collaborating.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fabricjs/fabric.js/pull/7657#issuecomment-1040223812>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIGAW4PSFK2MHM4GP4UBBHDU3JCMNANCNFSM5NVQVJBA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

asturur commented:
That is fine, when you are ok with this pr, you press the approve button.
i'll wait.

ShaMan123 reviewed (approved):
Eraser is committed
I've pushed some minor JSDOC changes as well
1. I want to change `enlivenObjects` not to use `getKlass` at all (we'll do that once we migrate js)
2. I do think we should consider wrapping the external methods of the svg parser with promises (and leave internals as is until you decide it's fate) in order to unify fabric. But it can wait.


Excellent job! This was a huge tedious effort.
Group is up!

melchiar reviewed (dismissed):
This is great. So much garbage removed!

asturur commented on src/mixins/canvas_serialization.mixin.js:
```diff
@@ -3,24 +3,23 @@ fabric.util.object.extend(fabric.StaticCanvas.prototype, /** @lends fabric.Stati
    *   // `object` = fabric.Object instance
    *   // ... do some stuff ...
+   * }).then((canvas) => {
+   *   ... your canvas callback
```
no we can change that too

asturur commented:
off i pushed away reviews because i addressed the text comment.

rhanmiano commented:
Hello, is this feature available already, and in which version? Thanks

ShaMan123 commented:
v5+

rhanmiano commented:
Hi @ShaMan123 , thanks for the response. I just have a problem. After upgrading from `4.6.0` to `5.2.1`. I get an error similar to [this](https://github.com/fabricjs/fabric.js/issues/7830) using `loadFromJSON(...).then`. After inspecting the `dist/fabric.js` the structure of `loadFromJSON` is still using the `fn(json, callback, reviver)`. Am I missing something? Thank you.

Please see image for reference.
![image](https://user-images.githubusercontent.com/31644925/160266062-1dcb1ed6-1cc7-4926-837f-894bf738a9b6.png)

![image](https://user-images.githubusercontent.com/31644925/160266049-97e9e77e-c28e-425e-8b09-8035c69e4b23.png)

ShaMan123 commented:
Did you try building fabric as suggested in the issue you referenced?

rhanmiano commented:
I tried rebuilding the fabric also, and the generated `fabric.min.js` still contains the structure of `fn(json, callback, reviver)` just the minified version. And still having the same issue.

ShaMan123 commented:
I see the version has somehow got mixed with outdated code.
I suggest you install fabric's latest version directly from github.
```bash
npm i --save github:fabricjs/fabric.js
```

rhanmiano commented:
Thanks @ShaMan123 , that solves the issue on my side. Will wait for the next stable patch.
